src/main/java/org/spongepowered/common/data/provider/block/state/BlockStateDataProvider.java

```diff
@@ -134,13 +135,17 @@ public static void register(final DataProviderRegistrator registrator) {
     }
 
     // @formatter:off
-    private static <T, V extends Comparable<V>> void registerProperty(final DataProviderRegistrator registrator, final Key<Value<T>> key, final Property<V> property) {
+    private static <T extends Comparable<T>, V extends Comparable<V>> void registerProperty(final DataProviderRegistrator registrator, final Key<Value<T>> key, final Property<V> property) {
 
         registrator.asImmutable(BlockState.class)
                 .create(key)
                     .supports(bs -> bs.getOptionalValue(property).isPresent())
-                    .get(bs -> (T) bs.getOptionalValue(property).orElse(null))
-                    .set((bs, v) -> bs.setValue(property, (V) v));
+                    .get(bs -> {
+                        return ((org.spongepowered.api.block.BlockState) bs).stateProperty((StateProperty<T>) property).orElse(null);
+                    })
+                    .set((bs, v) -> {
+                        return (BlockState) ((org.spongepowered.api.block.BlockState) bs).withStateProperty((StateProperty<T>) property, v).orElse(null);
+                    });
```

Comment:
```suggestion
                .get(bs -> ((org.spongepowered.api.block.BlockState) bs).stateProperty((StateProperty<T>) property).orElse(null))
                .set((bs, v) -> (BlockState) ((org.spongepowered.api.block.BlockState) bs).withStateProperty((StateProperty<T>) property, v).orElse((org.spongepowered.api.block.BlockState) bs));
```
We generally return the last `BlockState` if the offered value is invalid. The contract says that we return empty optional if the key is not supported.

Reply:
> We generally return the last `BlockState` if the offered value is invalid. The contract says that we return empty optional if the key is not supported.

Exactly, I missed that